xray/segment.go

```diff
@@ -286,34 +296,56 @@ func (seg *Segment) RemoveSubsegment(remove *Segment) bool {
 			seg.rawSubsegments = seg.rawSubsegments[:len(seg.rawSubsegments)-1]
 
 			if seg.ParentSegment != seg {
-				seg.ParentSegment.Lock()
-				defer seg.ParentSegment.Unlock()
+				seg.openSegments--
+				seg.Unlock()
+				unlock = false
+
+				atomic.AddUint32(&seg.ParentSegment.totalSubSegments, ^uint32(0))
+			}
+			if unlock {
+				seg.Unlock()
 			}
-			seg.ParentSegment.totalSubSegments--
-			seg.openSegments--
 			return true
 		}
 	}
+	seg.Unlock()
 	return false
 }
 
+func (seg *Segment) isOrphan() bool {
+	return seg.parent == nil || seg.Type == "subsegment"
+}
+
 func (seg *Segment) emit() {
 	seg.ParentSegment.GetConfiguration().Emitter.Emit(seg)
 }
 
-func (seg *Segment) handleContextDone() {
+func (seg *Segment)    handleContextDone() {
 	seg.Lock()
-	defer seg.Unlock()
-
 	seg.ContextDone = true
 	if !seg.InProgress && !seg.Emitted {
-		seg.flush()
+		s := seg
+		for {
+			if s.flush() {
```

Comment:
Could you explain why we create a for loop here and the logic behind this?

Reply:
Flushing performance emit operation. For given Segment/Subsegment, we check if we are able to flush? -> If it has parent don't flush, if it has no parent then flush (there are more condition, you can check inside flush()). 

If it doesn't flush, we traverse to it parent and check whether we are able to flush parent.

So the for loop traverse from current node to root segment and at each level it checks, whether the node is able to flush. If a node is valid and ready to flush, we flush entire subtree. 


For another comment why we have  removed `sub.parent = nil` in `beforeEmitSubsegment()` is the reason for this "new" flush logic which is bottom up traversal of `Segment` tree. If we set parent to nil for a subsegment, we will emit each subsegment for non-streaming mode inside `flush()`. Which shouldn't be the case. The SDK sends entire Segment document in non-streaming mode.

This won't affect streaming mode since, since the streaming mode function `StreamCompletedSubsegments` breaks a subtree on the length of subtree and not on parent nil check.

Reply:
Isn't line 327 - 339 exactly the same as what you added on segment.go `Close()` method? Is it possible to re-use the code? I'm not super familiar with the entire code base but I'm wondering why there are two places doing exactly the same thing.

Reply:
Sure, reusing the code is a good point (explained in later comment). The `flush()` logic is called from 2 places in the code. [Close()](https://github.com/aws/aws-xray-sdk-go/blob/master/xray/segment.go#L250) and [handleContextDone()](https://github.com/aws/aws-xray-sdk-go/blob/master/xray/segment.go#L310). For later function, the context may be cancelled by the framework for a given incoming request due to abrupt termination of the request.  For this scenario, we have a sweeping `Segment` logic inside [handleContextDone](https://github.com/aws/aws-xray-sdk-go/blob/master/xray/segment.go#L310). Hence, 2 places calling flush().